ts/src/coinmetro.ts

```diff
@@ -0,0 +1,1922 @@
+         */
+        await this.loadMarkets ();
+        const market = this.market (symbol);
+        const request = {
+            'pair': market['id'],
+            'timeframe': this.safeString (this.timeframes, timeframe, timeframe),
+        };
+        let until = undefined;
+        if (since !== undefined) {
+            if ((since < 0) || (!Number.isInteger (since))) {
+                throw new BadRequest (this.id + ' fetchOHLCV() param "since" must be an integer greater than 0 or equal to 0'); // the exchange returns 422 "No more than 1000 candles per call" for float or negative values
+            }
+            request['from'] = since;
+            if (limit !== undefined) {
+                const duration = this.parseTimeframe (timeframe) * 1000;
+                // todo: the exchange returns candles including the last one (with timestamp equals param 'to') should we substract 1 from duration?
+                until = this.sum (since, duration * (limit));
+            }
+        } else {
+            request['from'] = ':from'; // this endpoint doesn't accept empty from and to params (setting them into the value described in the documentation)
+        }
+        until = this.safeInteger2 (params, 'till', 'until', until);
+        if (until !== undefined) {
+            params = this.omit (params, [ 'till', 'until' ]);
+            request['to'] = until;
+        } else {
+            request['to'] = ':to';
+        }
+        const response = await this.publicGetExchangeCandlesPairTimeframeFromTo (this.extend (request, params));
+        //
+        //     {
+        //         "candleHistory": [
+        //             {
+        //                 "pair": "ETHUSDT",
+        //                 "timeframe": 86400000,
+        //                 "timestamp": 1697673600000,
+        //                 "c": 1567.4409353098604,
+        //                 "h": 1566.7514068472303,
+        //                 "l": 1549.4563666936847,
+        //                 "o": 1563.4490341395904,
+        //                 "v": 0
+        //             },
+        //             {
+        //                 "pair": "ETHUSDT",
+        //                 "timeframe": 86400000,
+        //                 "timestamp": 1697760000000,
+        //                 "c": 1603.7831363339324,
+        //                 "h": 1625.0356823666407,
+        //                 "l": 1565.4629390011505,
+        //                 "o": 1566.8387619426028,
+        //                 "v": 0
+        //             },
+        //             ...
+        //         ]
+        //     }
+        //
+        const candleHistory = this.safeValue (response, 'candleHistory', []);
+        return this.parseOHLCVs (candleHistory, market, timeframe, since, limit);
+    }
+
+    parseOHLCV (ohlcv, market: Market = undefined): OHLCV {
+        return [
+            this.safeNumber (ohlcv, 'timestamp'),
```

Comment:
timestamps should always be parsed with safeInteger

Reply:
My bad. I'll fix it.